### PR TITLE
Add fallback for multiple quality options

### PR DESCRIPTION
**Pull request by svenanderzen**
## Description
This PR adds the possibility to add multiple quality options as fallback alternatives instead of only one, as requested by Abios. It does so by reading a comma-separated list of quality options.

## Important files to review
`main.py`
`ui/settings_dialog.ui`

## Screenshots
quality_options

## See also
#168

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -343,7 +339,7 @@ def on_fail_add_stream(self, err, args):
             (title, text) = args
             QtWidgets.QMessageBox().warning(self, title, text)
         elif err == AddStreamError.DEFAULT_QUALITY_MISSING:
-            (stream_options, url, quality) = args
+            (stream_options, url) = args
```
Comment (svenanderzen) [inline main.py]: Couldn't really see where the `quality` was used, so I removed it. Let me know if I made a brain-fart @kaszim

File: constants.py [hunk 2]
```diff
@@ -25,7 +25,7 @@
 CONFIG_BUFFER_SIZE = 'buffer_size'
 CONFIG_DEFAULT_VALUES = {
     CONFIG_MUTE: False,
-    CONFIG_QUALITY: "best",
+    CONFIG_QUALITY: ["720p", "480p", "360p", "160p"],
```
Comment (mictab) [inline constants.py]: Maybe extract these to an enum?

File: main.py [hunk 3]
```diff
@@ -77,9 +77,9 @@ def setup_ui(self):
 
         self.ui.show()
 
-    def setup_videoframe(self, stream_url, stream_options, quality):
+    def setup_videoframe(self, stream_url, stream_options, stream_quality):
         """Sets up a videoframe and with the provided stream information."""
-        self.model.add_new_videoframe(stream_url, stream_options, quality)
+        self.model.add_new_videoframe(stream_url, stream_options, stream_quality)
```
Comment (mictab) [inline main.py]: Nice renaming 👍

File: main.py [hunk 4]
```diff
@@ -129,30 +129,27 @@ def show_settings(self):
             self.dialog.findChild(QtCore.QObject, RECORD_SETTINGS).setChecked(True)
         if cfg[CONFIG_MUTE]:
             self.dialog.findChild(QtCore.QObject, MUTE_SETTINGS).setChecked(True)
-        index = self.dialog.findChild(QtCore.QObject, QUALITY_SETTINGS).findText(cfg[CONFIG_QUALITY])
-        self.dialog.findChild(QtCore.QObject, QUALITY_SETTINGS).setCurrentIndex(index)
+        self.dialog.findChild(QtCore.QObject, QUALITY_SETTINGS).setText(", ".join(cfg[CONFIG_QUALITY]))
```
Comment (mictab) [inline main.py]: Maybe extract this to its own variable for clarity? Might not be entirely clear right now what this does. Talking about the string joining.

File: main.py [hunk 5]
```diff
@@ -129,30 +129,27 @@ def show_settings(self):
...
-        quality = str(self.dialog.findChild(QtCore.QObject, QUALITY_SETTINGS).currentText())
-        buffer_size = self.dialog.findChild(QtCore.QObject, BUFFER_SIZE) \
-            .value()
-        if self.dialog.findChild(QtCore.QObject, RECORD_SETTINGS).isChecked():
-            is_buffer = True
-        if self.dialog.findChild(QtCore.QObject, MUTE_SETTINGS).isChecked():
-            is_mute = True
+        """Generate a config from the provided settings."""
+        # Quality options should be provided as a comma-separated string.
+        quality_text = self.dialog.findChild(QtCore.QObject, QUALITY_SETTINGS).text()
```
Comment (mictab) [inline main.py]: Nice refactoring.

File: main.py [hunk 6]
```diff
@@ -129,30 +129,27 @@ def show_settings(self):
...
-        cfg[CONFIG_MUTE] = is_mute
-        cfg[CONFIG_QUALITY] = quality
-        cfg[CONFIG_BUFFER_SIZE] = buffer_size
+        cfg[CONFIG_QUALITY] = quality_options
+        cfg[CONFIG_BUFFER_STREAM] = self.dialog.findChild(QtCore.QObject, RECORD_SETTINGS).isChecked()
+        cfg[CONFIG_MUTE] = self.dialog.findChild(QtCore.QObject, MUTE_SETTINGS).isChecked()
+        cfg[CONFIG_BUFFER_SIZE] = self.dialog.findChild(QtCore.QObject, BUFFER_SIZE).value()
 
-        QtWidgets.QMessageBox.critical(self, "Caution", "Some of your changes may require a restart in order to take effect.")
+        QtWidgets.QMessageBox.critical(
```
Comment (mictab) [inline main.py]: Maybe it's better UX to tell the user to restart the application regardless? Now they might think that a config has taken effect when it actually requires a restart, which could lead to frustration since they don't know which configs require restart.

File: main.py [hunk 7]
```diff
@@ -192,10 +189,6 @@ def add_new_stream(self, stream_url=None, stream_quality=None):
             if not ok:
                 return
 
-        # Use default quality if not specified
```
Comment (mictab) [inline main.py]: Good.

File: main.py (hunk 6 above)
Comment (svenanderzen) [inline main.py]: Yeah sure, but perhaps we can add an issue and leave that for a separate PR? This way we can solve it more nicely by not requiring a restart på reloading the `cfg` variable in the `config.dump()` method.

File: main.py (hunk 4 above)
Comment (svenanderzen) [inline main.py]: Yeah sure 👍

File: constants.py (hunk 2 above)
Comment (svenanderzen) [inline constants.py]: Hmm, good thought but these are just the default values for the config and shouldn't be used anywhere else. Also there's probably more than 40+ stream options we would have to save in the enums to have enums for all quality options, and even then we're not sure if we have all because the streamlink plugins might decide to change... For example for many plugins the 720 option name is `720`, for others it's `720p` and some other `http_720p` or `why_did_we_let_the_plugin_writers_decide_what_name_to_give_the_quality_options_720`

So for consistency I recommend we leave all quality options as strings in the application.

File: main.py (hunk 6 above)
Comment (mictab) [inline main.py]: Doesn't matter to me 👍

File: main.py [hunk 8]
```diff
@@ -180,7 +181,7 @@ def import_streams_from_clipboard(self):
         for stream in streams:
             self.add_new_stream(stream_url=stream)
 
-    def add_new_stream(self, stream_url=None, stream_quality=None):
+    def add_new_stream(self, stream_url=None, stream_qualities=cfg[CONFIG_QUALITY]):
```
Comment (kaszim) [inline main.py]: This forces a restar. In my opinion it is better to have it dynamically use whatever is in the config.
